jena-arq/src/main/java/org/apache/jena/atlas/data/SortedDataBag.java

```diff
@@ -69,19 +69,81 @@
     
     protected final ThresholdPolicy<E> policy;
     protected final SerializationFactory<E> serializationFactory;
-    protected final Comparator<? super E> comparator;
+    protected final CanAbortComparator comparator;
     
     protected boolean finishedAdding = false;
     protected boolean spilled = false;
     protected boolean closed = false;
+    protected volatile boolean cancelled;
     
     public SortedDataBag(ThresholdPolicy<E> policy, SerializationFactory<E> serializerFactory, Comparator<? super E> comparator)
     {
         this.policy = policy;
         this.serializationFactory = serializerFactory;
-        this.comparator = comparator;
+        this.comparator = new CanAbortComparator(comparator);
     }
     
+    private final class CanAbortComparator implements Comparator<E> 
+    	{
+    	/**
+    	    The test for whether the sort has been cancelled is
+    	    performed every <code>cancelTestFrequency</code> comparisons.
+    	    This reduces the (presumed) overhead of access to a
+    	    volatile boolean.    	    
+    	*/
+    	static final int cancelTestFrequency = 10000;
+    	
+    	/**
+    	    Count of the number of times this comparator has been called.
+    	*/
+		int count = 0;
+		
+		final Comparator<? super E> baseComparator;
+		
+		public CanAbortComparator(Comparator<? super E> comparator) 
+			{
+			this.baseComparator = comparator;
+			}
+
+		@Override public int compare(E o1, E o2) 
+		{	
+			count += 1;
+			if (count % cancelTestFrequency == 0) 
+			{
+				if (cancelled) throw new AbandonSort();
+			}
+			return baseComparator.compare(o1, o2);
+		}
+		
+		/**
+		    Sort the array <code>e</code> using this comparator
+		 	with the additional ability to abort the sort.
+		*/
+		public boolean abortableSort(E[] e) {
+			try { Arrays.sort(e, this); }
```

Comment:
If the concern of this PR is being able to deal with really big piles of data, maybe it's worth adding a switch to use `Arrays.parallelSort(e, this)` here?